aten/src/ATen/Utils.h

```diff
@@ -51,7 +52,7 @@ static inline TensorImpl* checked_dense_tensor_unwrap(const Tensor& expr, const
 static inline std::vector<TensorImpl*> checked_dense_tensor_list_unwrap(ArrayRef<Tensor> tensors, const char * name, int pos, DeviceType device_type, ScalarType scalar_type) {
   std::vector<TensorImpl*> unwrapped;
   unwrapped.reserve(tensors.size());
-  for (unsigned int i = 0; i < tensors.size(); ++i) {
+  for (const auto i : c10::irange(tensors.size())) {
```

Comment:
@r-barnes I see below compile error for this line and that doesn't make much sense to me, could you please shed some light on this?

1 error detected in the compilation of "C:/Users/circleci/project/build/win_tmp/bin/.tmpCINHGA/tmpxft_000027dc_00000000-7_IndexKernel.cpp1.ii".
IndexKernel.cu
Retry attempt: 1
C:/Users/circleci/project/aten/src\ATen/Utils.h(55): error: no instance of overloaded function "c10::irange" matches the argument list
            argument types are: (size_t)

This error come from below build: [https://app.circleci.com/pipelines/github/pytorch/pytorch/285755/workflows/b0c1e315-c4c7-41c5-9705-6b35f7f09321/jobs/11568091](https://app.circleci.com/pipelines/github/pytorch/pytorch/285755/workflows/b0c1e315-c4c7-41c5-9705-6b35f7f09321/jobs/11568091)

Reply:
This error comes from CUDA-10.1 `nvcc.exe`compiler.

Reply:
That's good idea; I don't suppose you could experiment and find out? My intention was to produce a compile-time error if the user used a larger type for the start of the range than for the end of the range.

Since the iteration type is currently based on the end-of-range variable using a larger type for start-of-range presents the possibility of a lossy/dangerous downcast. The situation isn't liable to arise often, but it seemed better to be safe than sorry.

Reply:
@frank-dong-ms: You could also split these more complicated issues into a separate PR.

Reply:
std::is_integral<size_t> is correctly evaluated per my experiment. I changed the irange class to use enable_if instead of enable_if_t and these error disappears... The only difference is enable_if_t is only available after C++14 so that might be the cause?

Reply:
That could be the cause! I did all my testing on environments with C++14/C++17, so it's probable that I inadvertently targeted those environments.